capsules/extra/src/date_time.rs

```diff
@@ -0,0 +1,354 @@
+    ///                 -previous 5 store the hour
+    ///                 -previous 3 store the day_of_the_week
+    ///
+    /// The two u32 numbers are returned as a tuple
+
+    fn date_as_u32_tuple(&self, set_date: date_time::Date) -> Result<(u32, u32), ErrorCode> {
+        let month = self.month_as_u32(set_date.month);
+        let dotw = self.dotw_as_u32(set_date.day_of_week);
+
+        let date = set_date.year as u32 * (1 << 9) as u32
+            + month as u32 * (1 << 5) as u32
+            + set_date.day as u32;
+        let time = dotw as u32 * (1 << 17) as u32
+            + set_date.hour as u32 * (1 << 12) as u32
+            + set_date.minute as u32 * (1 << 6) as u32
+            + set_date.seconds as u32;
+
+        Ok((date as u32, time as u32))
+    }
+
+    fn call_driver(&self, command: DateTimeCommand, r2: usize, r3: usize) -> CommandReturn {
+        match command {
+            DateTimeCommand::ReadDateTime => {
+                let date_result = self.date_time.get_date_time();
+                match date_result {
+                    Result::Ok(()) => {
+                        self.in_progress.set(true);
+                        CommandReturn::success()
+                    }
+                    Result::Err(e) => CommandReturn::failure(e),
+                }
+            }
+            DateTimeCommand::SetDateTime => {
+                let date;
+
+                match self.date_from_u32_tuple(r2 as u32, r3 as u32) {
+                    Result::Ok(d) => date = d,
+                    Result::Err(e) => {
+                        return CommandReturn::failure(e);
+                    }
+                };
+
+                let get_date_result = self.date_time.set_date_time(date);
+
+                match get_date_result {
+                    Result::Ok(()) => {
+                        self.in_progress.set(true);
+                        CommandReturn::success()
+                    }
+                    Result::Err(e) => CommandReturn::failure(e),
+                }
+            }
+        }
+    }
+
+    fn enqueue_command(
+        &self,
+        command: DateTimeCommand,
+        year_month_dotm: u32,
+        dotw_hour_min_sec: u32,
+        appid: ProcessId,
+    ) -> CommandReturn {
+        if !self.in_progress.get() {
+            let grant_enter_res = self.apps.enter(appid, |app, _| {
+                app.subscribed = true;
+            });
+            match grant_enter_res {
+                Ok(()) => self
+                    .call_driver(
+                        command,
+                        year_month_dotm as usize,
+                        dotw_hour_min_sec as usize,
+                    )
+                    .into(),
+                Err(_e) => CommandReturn::failure(ErrorCode::FAIL),
+            }
+        } else {
+            let grant_enter_res = self.apps.enter(appid, |app, _| {
+                app.subscribed = true;
+            });
+
```

Comment:
Need to actually store the command here.

Reply:
Wait how does this work? If an app calls get time while a different app calls set time, how does the second app actually get the time?